### PR TITLE
fix: upgrade ckb-vm to fix snapshot behavior

### PR DESCRIPTION
**Pull request by driftluo**
### What problem does this PR solve?

chunk run with the snapshot unable to support the binary replacement behavior of exec, which will cause abnormal behavior after resume.

The upgraded VM supports backing up the current code(executable pages)  to the snapshot to fix the above problems. https://github.com/nervosnetwork/ckb-vm/pull/218

and https://github.com/nervosnetwork/ckb/pull/3176 should keep its behavior until hardfork activated

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: script/src/syscalls/load_cell_data.rs [hunk 1]
```diff
@@ -104,13 +107,15 @@ impl<'a, DL: CellDataProvider + 'a> LoadCellData<'a, DL> {
...
             .ok_or(VMError::Unexpected)?
             .slice((content_offset as usize)..(content_end as usize));
-        machine.memory_mut().init_pages(
-            addr,
-            memory_size,
-            FLAG_DIRTY | FLAG_EXECUTABLE | FLAG_FREEZED,
-            Some(data),
-            0,
-        )?;
+        // To be consistent with the mainnet, add this flag after hardfork
```
Comment (mohanson) [inline script/src/syscalls/load_cell_data.rs:110]: I made a mistake before. Any write operation to the ckb-vm memory will be marked as dirty, and the correct logic should be:

```rs
machine.memory_mut().init_pages(
            addr,
            memory_size,
            FLAG_EXECUTABLE | FLAG_FREEZED,
            Some(data),
            0,
        )?;
        if !self.dirty_flag_enable {
            let mut current_addr = addr;
            while current_addr < addr + memory_size {
                let page = current_addr / RISCV_PAGESIZE as u64;
                machine.memory_mut().clear_flag(page, FLAG_DIRTY)?;
                current_addr += RISCV_PAGESIZE as u64;
            }
        }
```

**Review by mohanson — CHANGES_REQUESTED** [review]

**Comment by quake** [comment]
Suggest to add unit test to cover this PR.

**Comment by yangby-cryptape** [comment]
I pushed #3178 to add more tests, I think that PR should be merged before this PR.

## Notice

#3178 doesn't fix the snapshot behavior.

## How to enable the test for resuming from snapshot?

This PR and my PR has conflicts, so resolved the conflicts at first.
- This PR add `true` for all `LoadCellData::new(..)` in tests, but I think test both `true` and `false` is better.
- Find `is_vm_version_1_and_syscalls_2_enabled` is a bit complicated, I think make a method is better than copied those code twice.

Then find the follow code in `script/src/verify/tests/ckb_latest/features_since_v2021.rs`:

```rust
    // TODO fix resume from snapshot
    let should_be_invalid_permission = script_version <= ScriptVersion::V1;
```

And change the above code as follow:

```diff
-    // TODO fix resume from snapshot
-    let should_be_invalid_permission = script_version <= ScriptVersion::V1;
+    let should_be_invalid_permission = script_version < ScriptVersion::V1;
```

Make sure all tests are passed at last.

**Review by yangby-cryptape — APPROVED** [review]

**Review by mohanson — APPROVED** [review]

**Review by zhangsoledad — APPROVED** [review]

**Comment by zhangsoledad** [comment]
bors r= yangby-cryptape,mohanson,zhangsoledad

**Pull request merged**